docs(v1.9.0): add CHANGELOG (#10968)
longhorn/longhorn: CHANGELOG/CHANGELOG-1.9.0.md

## Longhorn v1.9.0 Release Notes

This latest version of Longhorn introduces several features, enhancements, and bug fixes that are intended to improve system quality and the overall user experience. Highlights include new V2 Data Engine features, orphaned instance deletion, and offline replica rebuilding, recurring system backup and the observability of Longhorn resources.

The Longhorn team appreciates your contributions and anticipates receiving feedback regarding this release.

For more information about release-related terminology, see [Releases](https://github.com/longhorn/longhorn#releases).

## Removal

### Environment Check Script

The environment check script (`environment_check.sh`), which was deprecated in v1.7.0, has been removed from v1.9.0. Use the [Longhorn Command Line Tool](https://longhorn.io/docs/1.9.0/advanced-resources/longhornctl/) to check the Longhorn environment for potential issues.

### Orphan-Auto-Deletion Setting

> The `orphan-auto-deletion` setting has been replaced by `orphan-resource-auto-deletion` in v1.9.0. To replicate the previous behavior (`orphan-auto-deletion`), include `replica-data` in the `orphan-resource-auto-deletion` value.

Reviewer: Suggest emphasizing that the enabled setting will be automatically migrated to the new setting during the upgrade.

For more information, see [Orphan Data Cleanup](https://longhorn.io/docs/1.9.0/advanced-resources/data-cleanup/).

### Deprecated Fields in CRDs

Deprecated fields have been removed from the CRDs. For details, see [#6684](https://github.com/longhorn/longhorn/issues/6684).

## Deprecation & Incompatibilities

### `longhorn.io/v1beta1` API

> The `v1beta1` version of the Longhorn API is deprecated in v1.9.0 and will be removed in v1.10.0. During Longhorn system upgrades, custom resources using `longhorn.io/v1beta1` are automatically migrated to `longhorn.io/v1beta2`. Deprecated APIs are no longer served and may therefore cause unexpected or unwanted behavior. Avoid using longhorn.io/v1beta1 in new code and, if possible, rewrite existing code to exclude this version.

Reviewer: >rewrite existing code to exclude this version.

What does this mean? Did you mean for users who use this API version directly?

BTW, the API version is no longer supported. If users attempt to use it, what happens? I believe resource updates will not be accepted with errors. If it's unsupported, that should not mean deprecation, since it completely doesn't work starting from 1.9. We can just say it's unsupported. WDYT? 

Additionally, you can include the relevant ticket in this section for reference.

Author: > What does this mean? Did you mean for users who use this API version directly?

Yes, but this might no need to mention.

> BTW, the API version is no longer supported. If users attempt to use it, what happens? I believe resource updates will not be accepted with errors. If it's unsupported, that should not mean deprecation, since it completely doesn't work starting from 1.9. We can just say it's unsupported. WDYT?

v1beta1's `served` is marked as `false`, so, yes, we can simply say it is unsupported.

Reviewer: Sounds good.

- Before upgrading to v1.9.0:
  - Verify that backup targets are functioning properly.
  - Create full backups of all volumes that use a V2 backing image.
  - Detach and delete these volumes after the backups complete.
  - In the **Backing Image** page, save the specifications of all V2 backing images, including the name and the image source.
  - Delete all V2 backing images.
- After upgrading:
  - Recreate the V2 backing images using the same names and image sources.
  - Restore the volumes from your backups.

For more details, see [Issue #10805](https://github.com/longhorn/longhorn/issues/10805).

## Primary Highlights

### New V2 Data Engine Features

Although the V2 Data Engine is still considered an experimental feature in this release, the core functions have been significantly enhanced.

- [Support UBLK Frontend](https://github.com/longhorn/longhorn/issues/9719): Support for UBLK frontend in the V2 Data Engine, which allows for better performance and resource utilization.
- [Disaster Recovery Volumes](https://github.com/longhorn/longhorn/issues/6613): Introduces disaster recovery volume support in the V2 Data Engine for improved data protection.

> - [Auto-Salvage Volumes](https://github.com/longhorn/longhorn/issues/8430): Support auto-salvage volume support in the V2 Data Engine for improved volume recovery.

Reviewer: We can mention the following `offline replica rebuilding` works for V2 as well, and add a reference to the corresponding head section. Also, orphan.

> - [Auto-Salvage Volumes](https://github.com/longhorn/longhorn/issues/8430): Support auto-salvage volume support in the V2 Data Engine for improved volume recovery.

Author: Orphan doesn't work for v2 volume because the snapshot checksum busy issue that needs more time to investigate.

> - [Auto-Salvage Volumes](https://github.com/longhorn/longhorn/issues/8430): Support auto-salvage volume support in the V2 Data Engine for improved volume recovery.

Author: > These two seem 1.8 features instead?

Yes, forgot to remove them from important note.

> - [Auto-Salvage Volumes](https://github.com/longhorn/longhorn/issues/8430): Support auto-salvage volume support in the V2 Data Engine for improved volume recovery.

Reviewer: How about the offline replica rebuild? IIRC, it's also supported for v2?

- [Storage Network](https://github.com/longhorn/longhorn/issues/6450): Introduces support for storage networks in the V2 Data Engine to allow network segregation.

### Recurring System Backup

Starting with Longhorn v1.9.0, you can create a recurring job for system backup creation.

For more information, see [#6534](https://github.com/longhorn/longhorn/issues/6534)

[Documentation](https://longhorn.io/docs/1.9.0/advanced-resources/system-backup-restore/backup-longhorn-system/) | [GitHub Issue](https://github.com/longhorn/longhorn/issues/6534)

### Offline Replica Rebuilding

> Longhorn introduces offline replica rebuilding, a feature that allows degraded volumes to automatically recover replicas even while the volume is detached. This capability minimizes the need for manual recovery steps, accelerates restoration, and ensures high data availability.

Reviewer: Mention this setting, is disabled by default.

[Documentation](https://longhorn.io/docs/1.9.0/advanced-resources/rebuilding/offline-replica-rebuilding/) | [GitHub Issue](https://github.com/longhorn/longhorn/issues/8443)

### Orphaned Instance Deletion

> Longhorn supports the capability to track orphaned instances, which are leftover resources such as replicas or engines that are no longer associated with any active volume. These orphaned instances can accumulate due to unexpected failures or improper cleanup. To help maintain system performance and reduce unnecessary resource usage, Longhorn allows these instances to be removed either automatically through built-in cleanup routines or manually by the user via the UI or CLI.

Reviewer: Mention this setting is disabled by default.

## Longhorn v1.9.0 Release Notes

> This latest version of Longhorn introduces several features, enhancements, and bug fixes that are intended to improve system quality and the overall user experience. Highlights include new V2 Data Engine features, orphaned instance deletion, and offline replica rebuilding, recurring system backup and the observability of Longhorn resources.
> 
> The Longhorn team appreciates your contributions and anticipates receiving feedback regarding this release.
> 
> For more information about release-related terminology, see [Releases](https://github.com/longhorn/longhorn#releases).

Reviewer 2: ```suggestion
Longhorn v1.9.0 brings new features, enhancements, and bug fixes aimed at improving system stability and user experience. Key highlights include V2 Data Engine improvements, orphaned instance deletion, offline replica rebuilding, recurring system backups, and enhanced observability of Longhorn resources.

The Longhorn team appreciates your contributions and anticipates receiving feedback regarding this release.

For terminology and background on Longhorn releases, see [Releases](https://github.com/longhorn/longhorn#releases).
```

## Removal

### Environment Check Script

> The environment check script (`environment_check.sh`), which was deprecated in v1.7.0, has been removed from v1.9.0. Use the [Longhorn Command Line Tool](https://longhorn.io/docs/1.9.0/advanced-resources/longhornctl/) to check the Longhorn environment for potential issues.
> 
> ### Orphan-Auto-Deletion Setting
> 
> The `orphan-auto-deletion` setting has been replaced by `orphan-resource-auto-deletion` in v1.9.0. To replicate the previous behavior (`orphan-auto-deletion`), include `replica-data` in the `orphan-resource-auto-deletion` value. The original `orphan-auto-deletion` setting will be automatically migrated to `orphan-resource-auto-deletion` during the upgrade process.

Reviewer 2: ```suggestion
The `environment_check.sh` script, deprecated in v1.7.0, has been removed in v1.9.0. Use the [Longhorn Command Line Tool](https://longhorn.io/docs/1.9.0/advanced-resources/longhornctl/) (`longhornctl`) to check your environment for potential issues.

### Orphan Resource Auto-Deletion

The `orphan-auto-deletion` setting has been replaced by `orphan-resource-auto-deletion` in v1.9.0. To replicate the previous behavior, include `replica-data` in the `orphan-resource-auto-deletion` value. During the upgrade, the original `orphan-auto-deletion` setting is automatically migrated.
```

For more information, see [Orphan Data Cleanup](https://longhorn.io/docs/1.9.0/advanced-resources/data-cleanup/).

### Deprecated Fields in `longhorn.io/v1beta2` CRDs

> Deprecated fields have been removed from the CRDs. For details, see [#6684](https://github.com/longhorn/longhorn/issues/6684).

Reviewer 2: ```suggestion
Deprecated fields have been removed from the CRDs. For details, see issue [#6684](https://github.com/longhorn/longhorn/issues/6684).
```

### Breaking Change in V2 Backing Image

Starting with Longhorn v1.9.0, V2 backing images are incompatible with earlier versions due to naming conflicts in the extended attributes (`xattrs`) used by SPDK backing image logical volumes. As a result, V2 backing images must be deleted and recreated during the upgrade process. Since backing images cannot be deleted while volumes using them still exist, you must first back up, delete, and later restore those volumes as the following steps:

- Before upgrading to v1.9.0:
  - Verify that backup targets are functioning properly.
  - Create full backups of all volumes that use a V2 backing image.
  - Detach and delete these volumes after the backups complete.
  - In the **Backing Image** page, save the specifications of all V2 backing images, including the name and the image source.
  - Delete all V2 backing images.
- After upgrading:
  - Recreate the V2 backing images using the same names and image sources.
  - Restore the volumes from your backups.

For more details, see [Issue #10805](https://github.com/longhorn/longhorn/issues/10805).

## Primary Highlights

### New V2 Data Engine Features

> Although the V2 Data Engine is still considered an experimental feature in this release, the core functions have been significantly enhanced.

Reviewer 2: ```suggestion
While the V2 Data Engine remains experimental in this release, several core functions have been significantly improved:
```

### Recurring System Backup

Starting with Longhorn v1.9.0, you can create a recurring job for system backup creation.

For more information, see [#6534](https://github.com/longhorn/longhorn/issues/6534)

[Documentation](https://longhorn.io/docs/1.9.0/advanced-resources/system-backup-restore/backup-longhorn-system/) | [GitHub Issue](https://github.com/longhorn/longhorn/issues/6534)

### Offline Replica Rebuilding

Longhorn introduces offline replica rebuilding, a feature that allows degraded volumes to automatically recover replicas even while the volume is detached. This capability minimizes the need for manual recovery steps, accelerates restoration, and ensures high data availability. By default, offline replica rebuilding is disabled. To enable it, set the `offline-replica-rebuilding` setting to `true` in the Longhorn UI or CLI.

[Documentation](https://longhorn.io/docs/1.9.0/advanced-resources/rebuilding/offline-replica-rebuilding/) | [GitHub Issue](https://github.com/longhorn/longhorn/issues/8443)

### Orphaned Instance Deletion

> Longhorn supports the capability to track orphaned instances, which are leftover resources such as replicas or engines that are no longer associated with any active volume. These orphaned instances can accumulate due to unexpected failures or improper cleanup. To help maintain system performance and reduce unnecessary resource usage, Longhorn allows these instances to be removed either automatically through built-in cleanup routines or manually by the user via the UI or CLI. By default, the functionality is disabled. To enable it, set the `orphan-resource-auto-deletion` setting to `instance` in the Longhorn UI or CLI.

Reviewer 2: ```suggestion
Longhorn can now track and remove orphaned instances, which are leftover resources like replicas or engines that are no longer associated with an active volume. These instances may accumulate due to unexpected failures or incomplete cleanup.

To reduce resource usage and maintain system performance, Longhorn supports both automatic and manual cleanup. By default, this feature is disabled. To enable it, set the `orphan-resource-auto-deletion` setting to `instance` in the Longhorn UI or CLI.
```

[Documentation](https://longhorn.io/docs/1.9.0/advanced-resources/data-cleanup/orphaned-instance-cleanup/) | [GitHub Issue](https://github.com/longhorn/longhorn/issues/6764)

### Improved Metrics for Replica, Engine, and Rebuild Status

> Longhorn enhances observability by introducing new Prometheus metrics that expose the state and identity of Replica and Engine CRs, as well as the rebuild status. These improvements make it easier to monitor rebuild events across the entire cluster.

Reviewer 2: ```suggestion
Longhorn improves observability with new Prometheus metrics that expose the status and identity of Replica and Engine CRs, along with rebuild activity. These metrics make it easier to monitor rebuilds across the cluster.
```

- [DOC] Update the steps to set up the Azure backup target [9688](https://github.com/longhorn/longhorn/issues/9688) - @mantissahz
- [DOC] Explain the process of creating backing images from existing volumes [10093](https://github.com/longhorn/longhorn/issues/10093) - @ChanYiLin @chriscchien
- [DOC] Update ArgoCD installation document [10588](https://github.com/longhorn/longhorn/issues/10588) - @mantissahz
- [TASK] Remove environment check script since v1.9.0 [9239](https://github.com/longhorn/longhorn/issues/9239) - @yangchiu @derekbit
- [TASK] Add platform architect and volume encryption info to metrics [7047](https://github.com/longhorn/longhorn/issues/7047) - @c3y1huang @roger-ryao
- [DOC] Update the document and create a KB to address the limitation that BackingImage should be multiple of 512B [10536](https://github.com/longhorn/longhorn/issues/10536) - @ChanYiLin
- [DOC] Clarifications on defaultSettings.defaultDataLocality and persistence.defaultDataLocality usage [10253](https://github.com/longhorn/longhorn/issues/10253) - @james-munson @roger-ryao
- [DOC] Replica count behaviour is unclear [10861](https://github.com/longhorn/longhorn/issues/10861) - @hoo29 @chriscchien

> ## New Contributors

Reviewer: Like this!
